### ISSUE TITLE
Load testing exporter

### ISSUE THREAD

**Issue by xkilian** [IHI-Energy-Storage/sparkpluggw#8]
Have you done any load testing to indicate how many messages per second it supports?
The same for number of metrics that can be exposed by a single exporter instance?

**Comment by ober37** [comment]
Unfortunately not explicitly.

In production, we have roughly 500 metrics that we are gathering per second with this exporter running on a Dell R340 without issue.

**Comment by xkilian** [comment]
We are looking at a very similar number, so that should not be a problem. The only issue I see we will need to adress how can Prometheus scrape all the exposed metrics. As we will be exceeding the recommended maximum of 3000 Metrics exposed on a single http endpoint. How many distinct metrics are your exposing to prometheus? Do you think that the exporter could expose more than one metric endpoint, would you be open to a PR to permit? Or have other suggestions.

**Comment by ober37** [comment]
Yeah for our current edge nodes, I believe are maximum metrics (including unique label sets) are around 1000 on the high end.  I would definitely be open to a PR here and definitely would prefer this over a fork.

There needs to be a clean way to distinguish groups of metrics and what belongs to what metric endpoint.    One method to do this is to leverage the MQTT topic and allow for multiple MQTT listeners that are 1-1 with the metric endpoint.    Another method would be to leverage the various sp_XXX labels produced internally to allow for specific values (or combination of values) to route to a specific metric endpoint.    Which has the benefit of not forcing a specific MQTT topic methodology.

Thoughts??

**Comment by xkilian** [comment]
Most often the data stream is within the same topic, and I agree that that forcing users to use different topics dhould not be a requirement. I see more of a label based approach to split de stream of created metrics hashed to multiple web endpoints that can be gathered by Prometheus in parallel. So configuration options would be number of endpoints to spawn, starting port number for range, rule used to split the data stream to the various buffers that are to be exposed to each web endpoint. In terms of implementation, it could be a dedicated metric buffer per web endpoint. 
The other alternative is implementing remote write, to stream the data to Cortex instead of Prometheus. I see this as a future option for the exporter if required.

**Comment by xkilian** [comment]
Hey Brian, 

We will be looking at implementing remote-write to a Prometheus backend, now that Prometheus supports ingesting from remote-write sources. The design doc will consider the remote write(to Cortex) code from open-telemetry as its base to add to this exporter. I expect it to be a runtime flag to select if output is remote-write or a web endpoint. 
The hypothesis is that remote-write should be more resilient and performant than scraping for this type of use-case.

I also plan on adding recognition of events in the sparkplug data that would not be sent as metrics but would be sent as events to:

-  OPTION1: a local file buffer and then that would be processed by a syslogng/rsyslogd/fluentbit/etc to be then forwarded securely to a logging backend
- OPTION2: Directly as a structured syslog event with a go library that supports buffering/reconnection/etc to any logging backend
- OPTION3: promtail push API end up in Loki; (Not a fan of this as it is vendor locked)

After that, we might add some processing flexibility for parsing the incoming data that might or might not be released (if it makes sense for others).

And of course we will load test.. ;-)

What do you think? You can send me an email.

**Comment by ober37** [comment]
Sorry about the delay here, definitely like the ideas.  Here are my initial thoughts

1) On the remote-write option that makes perfect sense to me

2) I like the concept on the syslog integration, I am not sure what you mean by "recognition of events"?   Are you specifically talking about sparkplug events (e.g. like BIRTH events) or are you talking about some regex engine that would treat certain data as logging vs metrics?

3) I definitely would not be a fan of OPTION 3, but I think OPTION 1 or OPTION 2 make sense to me.    I think OPTION2 is probably the cleanest and simplest in my view but I would be curious to hear your thoughts here.

**Comment by xkilian** [comment]
1 - ok, super.
2 - The latter, a regex engine that will route on match to the logging output instead of the metrics output.

**Comment by xkilian** [comment]
Hello Brian, here are some updated news on the modifications of the plugin.

1. Remote-write and prometheus exposition format in a web page (as is today) are supported.
2. Logging is implemented and being sent using a loki library, so it will write directly to loki. This ensures that labels are consistent between metrics and events. I did indicate I was less of a fan for this, but Loki made sense in the overall design, that is not to say, that that it can't later be modified to add syslog output, but it would push the complexity of labelling out to a third party... And if we are using Prometheus, we are probably using Loki.
3. My collegue has also made improvements to a sparkplug generator, when it is ready we can link to it in the documentation.

He will contact you to go through the proposed changes.
